### PR TITLE
fix(server): associate ctx-scoped elicitation/sampling with the originating request (SEP-2260)

### PR DESCRIPTION
**Pull request by felixweinberger**
> 🗺️ **[2026-07-28 Spec Implementation — milestone tracker](https://gist.github.com/felixweinberger/722c541d7ed89069598ffb6ce64992e0)** · This PR: **M1: SEP-2260 association hardening**

Implements the 2025-lineage improvements from SEP-2260 (server-request association): ctx-scoped elicitation/sampling requests now ride the originating request's response stream, and the streamable HTTP transport gains an opt-in SSE keepalive for long human-in-the-loop waits. Adds the 4 e2e requirements that pin these behaviors.

## Motivation and Context

SEP-2260 (Final) requires server→client requests to be associated with the originating client request. Tracking issue: #2186.

Scope note from spec analysis: SEP-2260's MUST-strength rules never landed in the published 2025-11-25 spec pages (which remain SHOULD-strength: server→client messages "SHOULD relate to the originating client request"), and in the current draft spec they are enforced structurally by MRTR (SEP-2322) rather than as standalone rules. This PR therefore ships what is meaningful on the 2025 lineage:

1. **Association fix**: `ctx.mcpReq.elicitInput()` / `requestSampling()` were sent without `relatedRequestId`, so on streamable HTTP they were routed to the standalone GET stream instead of the originating POST response stream (and silently undeliverable when no GET stream was open). They now carry the handled request's id, matching `ctx.mcpReq.send()` semantics.
2. **SSE keepalive** (opt-in `keepAliveInterval` option): per SEP-2260's timeout considerations, transports SHOULD use SSE keepalives so connections survive long nested waits. Default off — no behavior change unless configured.

The remaining SEP-2260 rules (client rejects unsolicited server→client requests; no API for unsolicited pushes) assert draft-era semantics and land with the MRTR implementation (#2187), where they hold by construction.

## How Has This Been Tested?

4 new e2e requirements with scenarios (`protocol:assoc:*`), all passing:
- `nested-on-originating-stream` — nested elicit/sample arrives on the originating POST SSE stream
- `keepalive-during-nested` — keepalive comments observed during a nested wait
- `ping-exempt` — server-initiated ping remains valid outside any request context
- plus the association behavior under `enableJsonResponse` documented

Full e2e matrix: 33 files, 1,237 cells = 1,145 passed + 92 expected-fail, **0 unexpected failures**; no pre-existing cell changed state. Server unit tests 68/68. Typecheck/lint clean.

## Breaking Changes

None intended. Behavior change (improvement): ctx-scoped elicitation/sampling requests now arrive on the originating request's response stream instead of the standalone GET stream — this is the spec-recommended routing and also makes these requests work when no GET stream is open. Handler-supplied options still override. Caveat: under `enableJsonResponse` mode, nested ctx-scoped requests are not deliverable (same pre-existing limitation as `ctx.mcpReq.send()`).

ctx-scoped elicitation/sampling in `enableJsonResponse` mode (or after the response stream closed) now fails with a clear error instead of being silently dropped. Elicitation fundamentally requires a streaming response. The standalone GET stream is never used as a fallback for request-associated messages.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist

- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [ ] I have added or updated documentation as needed

## Additional context

- API shape of `keepAliveInterval` (name, ms unit, opt-in default) is open for review feedback.
- Part of the 2026-07-28 spec implementation plan (project 41); milestone M1.

### CODE REVIEW

**Comment by felixweinberger** [comment]
Review follow-up: added 12 unit tests in `@modelcontextprotocol/server` (association: relatedRequestId at the transport level, explicit-override precedence, top-level APIs unaffected; keepalive: timer lifecycle, cleanup on stream completion/transport close, concurrent streams). The tests caught one real edge case: `keepAliveInterval <= 0` was previously fed to `setInterval` (Node clamps to 1ms → keepalive flood); now treated as disabled.

**Comment by felixweinberger** [comment]
Updated per maintainer review: dropped the verbose section comment (tests are self-explanatory); eliminated the new non-null assertions by starting the keepalive inside the stream's `start` callback; `startKeepAlive` now always returns a stop function so cleanup sites call it unconditionally. Server tests 80/80, e2e 1,237 cells / 0 unexpected.

**Comment by felixweinberger** [comment]
Un-stacked: this PR has no dependency on #2226 or #2227 (zero file overlap; its requirement entries use no new manifest fields) — now based directly on main, mergeable independently.

File: packages/server/src/server/server.ts [hunk 1]
```diff
@@ -160,8 +160,11 @@ export class Server extends Protocol<ServerContext> {
             mcpReq: {
                 ...ctx.mcpReq,
                 log: (level, data, logger) => this.sendLoggingMessage({ level, data, logger }),
-                elicitInput: (params, options) => this.elicitInput(params, options),
-                requestSampling: (params, options) => this.createMessage(params, options)
+                // Associate the nested server→client request with the request being handled so
+                // transports can route it onto the originating request's response stream
+                // (servers must only send these requests in association with a client request).
+                elicitInput: (params, options) => this.elicitInput(params, { relatedRequestId: ctx.mcpReq.id, ...options }),
+                requestSampling: (params, options) => this.createMessage(params, { relatedRequestId: ctx.mcpReq.id, ...options })
```
[the remark this answers is not in the document]
Comment (felixweinberger) [reply packages/server/src/server/server.ts]: Fixed in 270945d0: `send()` now falls back to the standalone SSE stream for server→client requests whose related response stream cannot carry SSE (JSON-response mode, or the stream is gone) — restoring the routing these requests had before they carried a `relatedRequestId`. Unit test added: "delivers related server→client requests on the standalone GET stream".

File: test/e2e/requirements.ts [hunk 2]
```diff
@@ -787,6 +784,30 @@ export const REQUIREMENTS: Record<string, Requirement> = {
...
+        transports: ['streamableHttp'],
+        note: 'SHOULD-strength transport guidance (SEP Timeout Considerations); doubles as a regression guard against idle-stream drops during human-in-the-loop waits. This exercises the HTTP hosting layer; the matrix transport arg is ignored, so it runs as a single streamableHttp-labelled cell to avoid duplicate runs.'
+    },
+    'protocol:assoc:ping-exempt': {
+        source: 'https://modelcontextprotocol.io/seps/2260-Require-Server-requests-to-be-associated-with-Client-requests#1-add-warning-blocks-to-feature-documentation',
+        behavior:
+            '[SEP-2260] ping is exempt from the request-association rule: either party MAY send ping at any time on an established connection — including the server with no client request in flight — and receives an empty result.',
+        transports: STATEFUL_TRANSPORTS,
+        note: 'Server→client ping with nothing in flight needs a server→client channel; stateless hosting has none. Ping itself is removed in the 2026 draft, so this entry is 2025-lineage only.'
```
[the remark this answers is not in the document]
Comment (felixweinberger) [reply test/e2e/requirements.ts]: Fixed in 94125e73: dropped the lineage claim from the note. The manifest can't express that bound yet (no version-label support on this branch), and prose claiming a bound the data doesn't enforce is exactly the inconsistency you're pointing at. The retirement bound will be recorded when the draft-spec implementation makes it real.

File: packages/server/src/server/streamableHttp.ts [hunk 3]
```diff
@@ -564,6 +586,29 @@ export class WebStandardStreamableHTTPServerTransport implements Transport {
...
+        const timer = setInterval(() => {
+            try {
+                controller.enqueue(encoder.encode(': keepalive\n\n'));
+            } catch {
+                // Stream already closed or cancelled
+                clearInterval(timer);
+            }
+        }, this._keepAliveInterval);
+        return () => clearInterval(timer);
+    }
```
[the remark this answers is not in the document]
Comment (felixweinberger) [reply packages/server/src/server/streamableHttp.ts:612]: Fixed in 270945d0: the guard now treats any non-finite or non-positive interval as disabled — `interval === undefined || !Number.isFinite(interval) || interval <= 0` — covering NaN, Infinity, 0, and negatives, and returns the shared `noKeepAlive` constant. Unit tests added for NaN and Infinity.

File: packages/server/src/server/streamableHttp.ts [hunk 4]
```diff
@@ -141,6 +141,19 @@ export interface WebStandardStreamableHTTPServerTransportOptions {
...
+     * When set, the transport periodically writes an SSE comment (`: keepalive`) on every
+     * open SSE stream (POST response streams and the standalone GET stream) so idle
+     * connections are not torn down by intermediaries — for example while a request handler
+     * is waiting on a nested server→client request such as `elicitation/create`
+     * (human-in-the-loop delays).
+     *
+     * Disabled by default. Non-positive values (`<= 0`) are treated as disabled.
+     */
+    keepAliveInterval?: number;
+
```
[the remark this answers is not in the document]
Comment (felixweinberger) [reply packages/server/src/server/streamableHttp.ts:159]: Added in cd284ca9: `keepAliveInterval` is now documented in the streamable HTTP **Options:** prose in docs/server.md.

File: packages/server/src/server/streamableHttp.ts [hunk 5]
```diff
@@ -519,10 +540,12 @@ export class WebStandardStreamableHTTPServerTransport implements Transport {
...
             let streamController: ReadableStreamDefaultController<Uint8Array>;
+            let stopKeepAlive: () => void = noKeepAlive;
 
             const readable = new ReadableStream<Uint8Array>({
                 start: controller => {
                     streamController = controller;
+                    stopKeepAlive = this.startKeepAlive(controller, encoder);
                 },
                 cancel: () => {
                     // Stream was cancelled by client
```
[the remark this answers is not in the document]
Comment (felixweinberger) [reply packages/server/src/server/streamableHttp.ts]: Fixed in 270945d0: `stopKeepAlive` is declared outside the `try` and the catch block now stops the timer before returning the 500, so a failed replay can no longer leak the interval. The other two stream sites register their mapping entry synchronously after construction (no await in between), so the leak was replay-only. Unit test added: "stops the keepalive timer when event replay fails".

File: packages/server/src/server/streamableHttp.ts [hunk 6]
```diff
@@ -564,6 +588,26 @@ export class WebStandardStreamableHTTPServerTransport implements Transport {
...
 
+    /**
+     * Starts the SSE keepalive timer for a stream, writing comment frames until stopped.
+     * Returns a stop function, or `undefined` when keepalive is not configured or disabled
+     * (non-positive interval).
+     */
+    private startKeepAlive(controller: ReadableStreamDefaultController<Uint8Array>, encoder: InstanceType<typeof TextEncoder>): () => void {
+        if (this._keepAliveInterval === undefined || this._keepAliveInterval <= 0) {
+            return () => {};
+        }
```
[the remark this answers is not in the document]
Comment (felixweinberger) [reply packages/server/src/server/streamableHttp.ts:602]: Fixed in 270945d0: JSDoc reworded to match the actual contract (always returns a stop function; no-op when disabled), and the disabled branch returns the `noKeepAlive` constant instead of allocating a fresh closure.

**Comment by felixweinberger** [comment]
@claude review

**Comment by felixweinberger** [comment]
One more change from maintainer review (b979896): the ctx-scoped helpers now spread `options` **before** `relatedRequestId`, making the association non-overridable — `{ ...options, relatedRequestId: ctx.mcpReq.id }` instead of `{ relatedRequestId: ctx.mcpReq.id, ...options }`.

Rationale: SEP-2260 requires ctx-scoped server→client requests to be associated with the originating request; that's the entire purpose of the ctx helpers vs the top-level `Server` methods, so the association must not be overridable. This also fixes the gotcha where a caller passing an explicit `relatedRequestId: undefined` would silently clobber the default. Matches `ctx.mcpReq.send()`, which already does `{ ...options, relatedRequestId: request.id }`.

The "explicit override wins" unit test is inverted accordingly: it now asserts the association wins over a handler-supplied `relatedRequestId`.

File: packages/server/src/server/server.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (felixweinberger) [reply packages/server/src/server/server.ts]: Maintainer decision: fail-fast error instead of GET-stream fallback. Delivering unassociated requests violates the association rule this PR implements, and the standalone GET stream cannot be assumed to exist (clients are not required to open one). The regression concern is resolved by a deterministic error instead of a hang; elicitation in JSON-response mode was never functional in any meaningful sense.

File: docs/server.md [hunk 7]
```diff
@@ -48,7 +48,7 @@ const transport = new NodeStreamableHTTPServerTransport({
 await server.connect(transport);
 ```
 
-**Options:** Set `sessionIdGenerator` to a function (shown above) for stateful sessions. Set it to `undefined` for stateless mode (simpler, but does not support resumability). Set `enableJsonResponse: true` to return plain JSON instead of SSE streams.
+**Options:** Set `sessionIdGenerator` to a function (shown above) for stateful sessions. Set it to `undefined` for stateless mode (simpler, but does not support resumability). Set `enableJsonResponse: true` to return plain JSON instead of SSE streams. Set `keepAliveInterval` (milliseconds) to write periodic SSE keepalive comments on open streams, so proxies and load balancers do not drop idle connections during long-running requests.
```
[the remark this answers is not in the document]
Comment (felixweinberger) [reply docs/server.md]: Added — the Options line in docs/server.md now states that ctx-scoped server-initiated requests require an SSE response stream and throw under `enableJsonResponse: true`, and the Server-initiated requests section intro carries the same constraint. A changeset was added in the same commit covering the association change, the fail-fast behavior, and the new `keepAliveInterval` option.

**Comment by felixweinberger** [comment]
Closing this as will submit as a bigger unified one implementing 2575+2567 e2e _without_ backchannel initially.
